osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/EffectiveBPMLoader.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.ControlPoints;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing
+{
+    public class EffectiveBPMLoader
+    {
+        private readonly IList<TaikoDifficultyHitObject> noteObjects;
+        private readonly IReadOnlyList<TimingControlPoint?> timingControlPoints;
+        private readonly double globalSliderVelocity;
+
+        public EffectiveBPMLoader(IBeatmap beatmap, List<TaikoDifficultyHitObject> noteObjects)
+        {
+            this.noteObjects = noteObjects;
+            timingControlPoints = beatmap.ControlPointInfo.TimingPoints;
+            globalSliderVelocity = beatmap.Difficulty.SliderMultiplier;
+        }
+
+        /// <summary>
+        /// Calculates and sets the effective BPM and slider velocity for each note object, considering clock rate and scroll speed.
+        /// </summary>
+        public void LoadEffectiveBPM(ControlPointInfo controlPointInfo, double clockRate)
+        {
+            using var controlPointEnumerator = timingControlPoints.GetEnumerator();
+            controlPointEnumerator.MoveNext();
+            var currentControlPoint = controlPointEnumerator.Current;
```

Comment:
Why not `controlPointInfo.TimingPointAt(currentNoteObject.StartTime)` instead of this current/next search?

Reply:
I don't think the performance impact is noticable since it uses binary search anyway, and the increased readability would be more important imo.
Also why would sequential state have to be maintained? If, for some reason, the input note objects where out of order, shouldnt it still use the correct controlpoints instead of the latest one?

Reply:
Fair call. much simpler and just tested performance on a large set, pretty negligible. Main idea around sequential state was to leverage performance because its based on time, but yeah seems fine to change.